doc/E2_guidelines.md

```diff
@@ -1,96 +1,344 @@
+official standard but is widely supported by compilers and much easier to
+maintain.
 
-Once the code on the branch you are working on is ready (i.e. clean, tested, documented, formatted and with demos), you request for it to be merged with the master branch:
+Note that there are some border cases where `#pragma once` is causing issues
+(e.g. on Windows or when having a weird build setup with symlinks or copies of
+files). In such cases use traditional include guards. Make sure they have unique
+names by composing them from the package name and the path/name of the file
+(e.g. `MY_PACKAGE_PATH_TO_FILE_FILENAME_H`). Please **do not** add underscores
+as prefix or suffix because this is reserved for the compiler pre-proccesor
+variables.
 
-- visit the repository on github
+##### Keep scopes small
 
-- click on the link "Pull requests"
+Avoid adding anything to the global namespace if possible. This means
 
-- create a new pull request, asign yourself and select reviewers using the web interface
+- Use a namespace when defining extern symbols.
+- Use an anonymous namespace or static for symbols that are only used internally.
 
+Generally define symbols in the smallest possible scope, i.e. if a variable is
+only used inside one loop, define it inside this loop (however, do not consider
+this to be a very strict rule, deviate from it where it seems reasonable).
+
+##### Use types with explicit sizes
+
+The header *stdint* defines primitive types with explicit sizes:
+*int32_t*, *uint32_t*, *int16_t*, ...
+They should be preferred over the build-in types int, unsigned, short, ...
+To use them add the following include:
+
+    #include <stdint>
+
+
+## Merging to master/main branch
+
+Once the code on the branch you are working on is ready, you request for it to
+be merged with the master branch.
+
+
+### Before opening a pull request
+
+Before you create a pull request (or at least before you assign reviewers):
+- Rebase your branch on the latest master/main to make sure there have been no
+  conflicting changes in the meantime.
+- Remove any debug code or commented code that is not needed anymore.
+- Test all new features and also existing ones if they are affected by your
+  change.  Unit tests are great for this!
+- Document new features and update existing documentation if something changed.
+- Add new unit tests and demos where appropriate.
+- Run the auto-formatting tools on all new/modified files (see above).
+
+### Creating a pull request
+
+- Visit the repository on GitHub,
+- click on the link "Pull requests",
+- create a new pull request, assign yourself and select reviewers using the web
+  interface,
 - select an appropriate label.
 
 It is good manner to squash all the commits of the branch before a code review (see below), as this makes the work of the reviewer easier. See for example these [instructions](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html).
```

Comment:
How does this actually make reviewing easier?  I would rather suggest that sometimes it can help to have separate commits that can be reviewed one by one (however, in this case, it should probably rather be separate PRs).

Reply:
Ideally one push request should be about the change/improvement of only one feature, so having only one commit would make sense. 
On my side I typically push on the branch on my desktop in order to keep working on my laptop (or the opposite), and having related commit is just noise for the reviewer.
But maybe that should be more of a case by case issue that a general rule.
We can rephrase:

"When relevant, it is good manner ..."

Reply:
I see your point.  How about this?

```suggestion
It is good manner to squash commits that logically belong together (e.g. multiple small changes on the same part of code) before a code review (see below), as this makes the work of the reviewer easier and leads to a cleaner git history. See for example these [instructions](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html).
```

Reply:
+1